pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BatchMessageWithBatchIndexLevelTest.java

```diff
@@ -0,0 +1,104 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import static org.testng.Assert.assertEquals;
+
+@Test(groups = "broker")
+public class BatchMessageWithBatchIndexLevelTest extends BatchMessageTest {
+
+    @BeforeClass
+    @Override
+    protected void setup() throws Exception {
+        conf.setAcknowledgmentAtBatchIndexLevelEnabled(true);
+        super.baseSetup();
+    }
+
+    @Test
+    @SneakyThrows
+    public void testBatchMessageAck() {
+        int numMsgs = 40;
+        final String topicName = "persistent://prop/ns-abc/batchMessageAck-" + UUID.randomUUID();
+        final String subscriptionName = "sub-batch-1";
+
+        ConsumerImpl<byte[]> consumer = (ConsumerImpl<byte[]>) pulsarClient
+                .newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .receiverQueueSize(10)
+                .subscriptionType(SubscriptionType.Shared)
+                .enableBatchIndexAcknowledgment(true)
+                .negativeAckRedeliveryDelay(100, TimeUnit.MILLISECONDS)
+                .subscribe();
+
+        Producer<byte[]> producer = pulsarClient
+                .newProducer()
+                .topic(topicName)
+                .batchingMaxMessages(20)
+                .batchingMaxPublishDelay(200, TimeUnit.MILLISECONDS)
+                .enableBatching(true)
+                .create();
+
+        List<CompletableFuture<MessageId>> sendFutureList = Lists.newArrayList();
+        for (int i = 0; i < numMsgs; i++) {
+            byte[] message = ("batch-message-" + i).getBytes();
+            sendFutureList.add(producer.newMessage().value(message).sendAsync());
+        }
+        FutureUtil.waitForAll(sendFutureList).get();
+        PersistentTopic topic = (PersistentTopic) pulsar.getBrokerService().getTopicReference(topicName).get();
+        PersistentDispatcherMultipleConsumers dispatcher = (PersistentDispatcherMultipleConsumers) topic
+                .getSubscription(subscriptionName).getDispatcher();
+        Message<byte[]> receive1 = consumer.receive();
+        Message<byte[]> receive2 = consumer.receive();
+        consumer.acknowledge(receive1);
+        consumer.acknowledge(receive2);
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(dispatcher.getConsumers().get(0).getUnackedMessages(), 18);
+        });
+        Message<byte[]> receive3 = consumer.receive();
+        Message<byte[]> receive4 = consumer.receive();
+        consumer.acknowledge(receive3);
+        consumer.acknowledge(receive4);
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(dispatcher.getConsumers().get(0).getUnackedMessages(), 16);
+        });
+        Message<byte[]> receive5 = consumer.receive();
+        consumer.negativeAcknowledge(receive5);
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(dispatcher.getConsumers().get(0).getUnackedMessages(), 16);
```

Comment:
the value is already 16, so...what are we testing here ?

Reply:
Here, we call `consumer.negativeAcknowledge(receive5);` 
As `negativeAcknowledge` will mapped to Consumer#redeliverUnacknowledgedMessages which will calculate unack-msg.  so here, I will test the unack-msg count is right.

Reply:
but you are not verifying that the number is something different from 16 before this assertion.

I see two cases:
1) you expect that the value is 16 at line 96, then it changes to another value, then is it again back to 16
2) you are expecting that the value is never changing

if it is 1) then you should add some assertion between this line and like 96 to check that the value is not 16 anymore
if it is case 2) then you don't need `Awaitility.await().untilAsserted` but you can simply leave
`assertEquals(dispatcher.getConsumers().get(0).getUnackedMessages(), 16);`

Reply:
When call `consumer.negativeAcknowledge(receive5)`, broker will execute `Consumer#redeliverUnacknowledgedMessages`. 
In `redeliverUnacknowledgedMessages`, I have added some logic about calculating the unack-msg. Please ref here : 
Consumer#line 875~880. I think you miss here.
As `negativeAcknowledge`  will be executed async, Then, I have to make the unack-msg count is also 16. So it should add `Awaitility.await()`

Reply:
@eolivelli

Reply:
> As negativeAcknowledge will be executed async, Then, I have to make the unack-msg count is also 16. So it should add Awaitility.await()

I agree, but how can you verify that the value simply **never changed** from 16 ?
because even if the code you changed did not run at all, then the value was 16 and it will still be 16, so your test is not adding any additional verification.

In other words: this test is not testing that you added the code, it may "work" even if your code is not executed because the message did not reach to the broker, because it is async.
we should add something that asserts that we are executing that code in Consumer#line 875~880

Reply:
Aha, I get your point.  Sorry for this.  
And as your reminder, I have checked Consumer#redeliverUnacknowledgedMessages, there is a wrong logic before, I have fixed and updated the test, please help review. Thanks again.

Reply:
@eolivelli

Reply:
so we found a problem in the patch by enhancing the test ?
I am happy that my comments helped in getting to a better patch.

Reply:
> so we found a problem in the patch by enhancing the test ? I am happy that my comments helped in getting to a better patch.

Thanks very much.